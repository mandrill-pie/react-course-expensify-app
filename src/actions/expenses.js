import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
	type: 'ADD_EXPENSE',
	expense
});

// Async action
export const startAddExpense = (expenseData = {}) => {
	return (dispatch) => {
		// Destructure received data (expenseData)
		const {
			description = '',
			note = '',
			amount = 0,
			createdAt = 0
		} = expenseData;

		// Save to db
		const expense = {
			description,
			note,
			amount,
			createdAt
		};
		return database.ref('expenses')
			.push(expense)
			.then((ref) => {
				// Update the Redux Store
				dispatch(addExpense({
					id: ref.key,
					...expense
				}));
			});
	};
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// Async action
export const startRemoveExpense = ({ id }) => {
	return (dispatch) => {
		// remove expense from DB
		return database.ref(`expenses/${id}`).remove().then(() => {
			// remove expense from state
			dispatch(removeExpense({ id }));
		});
	};
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});

// Async action
export const startEditExpense = (id, updates) => {
	return (dispatch) => {
		// Update expense in DB
		return database.ref(`expenses/${id}`)
			.update(updates)
			.then(() => {
				// Update the Redux Store
				dispatch(editExpense(id, updates));	
			})
			.catch((error) => {
				console.log('Error: ', error);
			});
	};
};

// SET_EXPENSES (get expenses from DB)
export const setExpenses = (expenses) => ({
	type: 'SET_EXPENSES',
	expenses
});

// Async action
export const startSetExpenses = () => {
	return (dispatch) => {
		// Get expenses from DB
		return database.ref('expenses')
			.once('value')
			.then((snapshot) => {
				const expenses = [];
				snapshot.forEach((childSnapshot) => {
					expenses.push({
						...childSnapshot.val(),
						id: childSnapshot.key,
					});
				});

				// Update the Redux Store
				dispatch(setExpenses(expenses));	
			})
			.catch((error) => {
				console.log('Error: ', error);
			});
	};
};