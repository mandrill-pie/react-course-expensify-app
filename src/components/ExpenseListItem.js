import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id, description, amount, createdAt}) => (
	<Link className="list-item" to={`/edit/${id}`}>
		<div>
			<h3 className="list-item__title">{description}</h3>
			<span className="list-item__date">{moment(createdAt).format('YYYY-MM-DD')}</span>
		</div>
		<div>
			<h3 className="list-item__amount">
				{numeral(amount).format('$0,0.00')}
			</h3>
		</div>
	</Link>
);

export default ExpenseListItem;