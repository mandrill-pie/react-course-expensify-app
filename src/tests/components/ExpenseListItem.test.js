import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('ExpenseListItem - render 1 list', () => {
	const expense = expenses[2];
	const wrapper = shallow(<ExpenseListItem {...expense} />);
	expect(wrapper).toMatchSnapshot();
});