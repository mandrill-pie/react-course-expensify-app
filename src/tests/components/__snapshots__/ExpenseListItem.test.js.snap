// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem - render 1 list 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/0f0gj3bouwouf"
>
  <div>
    <h3
      className="list-item__title"
    >
      Rented car
    </h3>
    <span
      className="list-item__date"
    >
      2017-10-05
    </span>
  </div>
  <div>
    <h3
      className="list-item__amount"
    >
      $143.16
    </h3>
  </div>
</Link>
`;
